### PR TITLE
Complete allocation utility Daml test coverage

### PR DESCRIPTION
**Pull request by meiersi-da**
See V2_VALIDATION.md changes for the changes and the few fixes contributed by this patch.

Fixes #4517

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If an upgrade test is required, comment `/upgrade_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a logical synchronizer upgrade test is required (from canton-3.5), comment `/lsu_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/canton-network/splice/blob/main/docs/src/release_notes.rst).
- [ ] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/canton-network/splice/blob/main/TESTING.md#running-and-debugging-integration-tests) or use your favorite screenshot tool

#### Merge Guidelines
- [ ] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

### CODE REVIEW

File: token-standard/examples/splice-token-test-trading-app-v2/daml/Splice/Testing/Apps/TradingAppV2.daml [hunk 1]
```diff
@@ -130,21 +130,24 @@ template OTCTrade with
...
-              (admin, legs) <- Map.toList $ splitLegsBy (._1) authorizerLegs
-              pure V2.AllocationSpecification with
-                admin
-                authorizer
-                transferLegSides = map (._2) legs
-                settlementDeadline
-                nextIterationFunding = None -- Note: this trading app does not make use of iterated settlement
-                committed = False
-                meta = emptyMetadata
+        results <- forA (Map.toList $ splitTradeLegsByAuthorizer tradeLegs) $ \(authorizer, authorizerLegs) -> do
```
Comment (meiersi-da) [inline token-standard/examples/splice-token-test-trading-app-v2/daml/Splice/Testing/Apps/TradingAppV2.daml:133]: stop creating alloc request for the venue

File: token-standard/splice-token-standard-utils/daml/Splice/TokenStandard/Utils/Internal/Allocations.daml [hunk 2]
```diff
@@ -244,21 +237,24 @@ ensureIsReceiptAllocation
...
-  -- restrict to single actor choices with the receiver as the actor
-  requireMatchExpected ("choiceArg.actors", actors) [receiver]
-  -- require all transfer legs to be receipts
-  forA_ transferLegSides $ \leg -> do
-    require ("transferLeg[" <> leg.transferLegId <> "] is a receipt leg") (leg.side == AllocationV2.ReceiverSide)
-  -- no holdings for funding
-  require' ("choiceArg.inputHoldingCids", inputHoldingCids) isEqualR ("emptyList", [])
-  -- no extra metadata to avoid accidental delegation over special functionality
-  require' ("choiceArg.extraArgs.meta", meta) isEqualR ("emptyMetadata", emptyMetadata)
+  case choiceArg of
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-utils/daml/Splice/TokenStandard/Utils/Internal/Allocations.daml:240]: switched to case and positional arguments to be one safe side wrt checking all args. I forgot to test the `nextIterationFunding` when I added it.

File: token-standard/splice-token-standard-v2-test/daml/Splice/Testing/Apps/TradingAppV2_Backend.daml [hunk 3]
```diff
@@ -126,11 +126,13 @@ querySettleableTrades registries venue = do
...
-
-    if minimallyRequiredAuthorizations `Set.isSubsetOf` actualAuthorizations
+    let missingAuthorizations = minimallyRequiredAuthorizations `Set.difference` actualAuthorizations
+    if Set.null missingAuthorizations
       then do
         Some <$> mkSettleableTrade trade agreementsByTrader
-      else
+      else do
+        debug "missingAuthorizations"
+        forA_ missingAuthorizations debug
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Testing/Apps/TradingAppV2_Backend.daml:135]: simplified debugging

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 4]
```diff
@@ -363,21 +409,132 @@ multiDvPWithFeeTest_onlyV2_cancel = runTest $ \env@TestEnvironment{..} ->
       , mkH X2 bob 15.0, mkH X2 bob 5.0, mkH Y2 bob 0.3, mkH Y2 bob 2.7
       ]
 
--- TODO(#4517): test withdrawal with an expired and one unexpired lock
-
--- TODO(#4517): add negative tests showing that bob must approve their forwarding allocation
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:368]: The comment here was about ensuring that the venue cannot just use their trading agreement to receive funds in the name of bob and forward them to themselves. This is not tested as part of `TestEnsureReceiptAllocation` where we show that even self-transfers are not considered receipt allocations.

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 5]
```diff
@@ -363,21 +409,132 @@ multiDvPWithFeeTest_onlyV2_cancel = runTest $ \env@TestEnvironment{..} ->
...
-
--- TODO(#4517): add cancellation/withdrawal and mixed version settlement test variants for this scenario as well
-
--- TODO(#4517): add the example from the CIP, and test for corner cases like cyclic trades that fully net with the max Decimal amount
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:372]: test for cyclic trades is part of CIP, max amount is tested as part of `test_settleBatchV2_decimal_overflow` in `TestAmuletAllocationNegative`

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 6]
```diff
@@ -363,21 +409,132 @@ multiDvPWithFeeTest_onlyV2_cancel = runTest $ \env@TestEnvironment{..} ->
...
-
--- TODO(#4517): test that duplicate allocations of the same leg cannot be settled
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:374]: tested by the existing `test_settleBatchV2_duplicate_*` tests in `TestAmuletAllocationNegative`

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 7]
```diff
@@ -363,21 +409,132 @@ multiDvPWithFeeTest_onlyV2_cancel = runTest $ \env@TestEnvironment{..} ->
...
+    expectedNumAllocationsV1 = 1  -- allocation of bob for bobToAliceX1
+    expectedNumAllocationsV2 = 4  -- bob (to venue AM, alice X2, alice Y2), alice (AM send)
+    expectedFinalHoldings =
+      [ mkH AM alice 20.0, mkH AM alice 80.0
+      , mkH AM bob 99.0, mkH AM bob 1.0
+      , mkH X1 bob 10.0, mkH X1 bob 40.0
+      , mkH X2 bob 15.0, mkH X2 bob 5.0, mkH Y2 bob 0.3, mkH Y2 bob 2.7
+      ]
 
--- TODO(#4515): test allocation workflow with different accounts for cash and security leg
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:376]: Tested by the change to receive some fees into a different account in the venue

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 8]
```diff
@@ -363,21 +409,132 @@ multiDvPWithFeeTest_onlyV2_cancel = runTest $ \env@TestEnvironment{..} ->
...
+    expectedNumAllocationsV1 = 1  -- allocation of bob for bobToAliceX1
+    expectedNumAllocationsV2 = 4  -- bob (to venue AM, alice X2, alice Y2), alice (AM send)
+    expectedFinalHoldings =
+      [ mkH AM alice 20.0, mkH AM alice 80.0
+      , mkH AM bob 99.0, mkH AM bob 1.0
+      , mkH X1 bob 10.0, mkH X1 bob 40.0
+      , mkH X2 bob 15.0, mkH X2 bob 5.0, mkH Y2 bob 0.3, mkH Y2 bob 2.7
+      ]
 
--- TODO(#4515): test V2 wallet accepting allocation request from a V1 app using mix of V1 and V2 assets
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:378]: tested in `TestSettlement_V1App`

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 9]
```diff
@@ -524,8 +687,6 @@ runTest mkTest = do
   -- check final holdings
   checkHoldings env ("holdings after " <> show test.testWorkflow) test.expectedFinalHoldings
 
-  -- TODO(#4517): check that featured app reward coupons get created, if that is still required
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:527]: left-out as they are going to be removed at about the same time, and guidance is anyways to post-create markers for traffic

File: token-standard/splice-token-standard-v2-test/daml.yaml [hunk 10]
```diff
@@ -36,7 +36,6 @@ data-dependencies:
   - ../splice-api-token-allocation-request-v2/.daml/dist/splice-api-token-allocation-request-v2-current.dar
   - ../splice-api-token-allocation-instruction-v2/.daml/dist/splice-api-token-allocation-instruction-v2-current.dar
 
-  # TODO(#4878): rename splice-token-standard-util to splice-util-token-standard-core to better align with the existing 'splice-util-token-standard-wallet'
```
Comment (meiersi-da) [inline token-standard/splice-token-standard-v2-test/daml.yaml:39]: left-over TODO. Won't do the change.

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml [hunk 11]
```diff
@@ -185,40 +184,46 @@ bilateralDvPWithFeeTest TestEnvironment{..} =
...
+      , bobToVenueX2
       ]
     expectedNumRequests = 3
     expectedNumAllocationsV1 = 0
     expectedNumAllocationsV2 = 4 -- alice, bob for both admins, venue does not allocate
     expectedHoldingsAfterAllocation =
       [ mkLH AM alice 20.0, mkH AM alice 80.0
       , mkH AM bob 100.0
-      , mkH X2 bob 10.0, mkLH X2 bob 40.0
+      , mkH X2 bob 8.0, mkLH X2 bob 42.0
```
Comment (moritzkiefer-da) [inline token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:197]: why do the numbers change here?

**Review by moritzkiefer-da — APPROVED** [review]
thanks!

File: token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml (hunk 11 above)
Comment (meiersi-da) [reply token-standard/splice-token-standard-v2-test/daml/Splice/Tests/TestSettlement_V1V2Mixed.daml:197]: because I added the additional X2 leg `bobToVenueX2` for 2.0 X2 that I'm using to test an allocation with a non-standard account for the venue as the target
